File: Rombouts.py
# --------------------------------------------------------------------------- #
# ROMBOUTS et al. 2015 VEGETATION MODEL
# --------------------------------------------------------------------------- #

from PyDSTool import *
DSargs = args(name='Rombouts_Vegetation_model')

# --------------------------------------------------------------------------- #
# PARAMETERS
# --------------------------------------------------------------------------- #

pars = {'Ct':500,'Q0':342.5,'p':0.3,'av':0.1,'ag':0.4,'alphamax':0.85,'alphamin':0.25,
'Tal':263,'Tau':300,'B0':200,'B1':2.5,'Topt':283,'k':0.004,'gmma':0.01}
DSargs.pars = pars

# --------------------------------------------------------------------------- #
# INITIAL CONDITIONS
# --------------------------------------------------------------------------- #

icdict = {'T':280,'A':0.9}
DSargs.ics = icdict
DSargs.tdomain = [0,100]

# --------------------------------------------------------------------------- #
# EQUATIONS
# --------------------------------------------------------------------------- #

dTstr = '((1-alpha(T,A))*Q0-R0(T))/Ct'
dAstr = 'beta(T)*A*(1-A)-gmma*A'
DSargs.varspecs = {'T': dTstr, 'A': dAstr}
DSargs.fnspecs  = {'alpha': (['T','A'], '(1-p)*alpha0(T)+p*(av*A+ag*(1-A))'),
                   'beta': (['T'],'max(0,1-k*(T-Topt)**2)'),
                   'alpha0': (['T'],'max(alpha01(T),alpha02(T))'),
                   'alpha01': (['T'],'alphamax+min(0,(alphamin-alphamax)*(T-Tal)/(Tau-Tal))'),
                   'alpha02': (['T'],'alphamin'),
                   'R0':(['T'],'B0+B1*(T-Topt)')}
                               
# --------------------------------------------------------------------------- #
# SETUP GENERATOR
# --------------------------------------------------------------------------- #
ode = Generator.Vode_ODEsystem(DSargs)
ode.set(pars = {'gmma': 0.01} )
ode.set(ics = {'T': 290,'A':0.1} )

# --------------------------------------------------------------------------- #
# RUN FORWARD
# --------------------------------------------------------------------------- #

print('Computing curve...')
start = clock()
traj = ode.compute('test_traj')
print('done in %.3f seconds!' % (clock()-start))
pts = traj.sample()

fig, (ax1, ax2)= plt.subplots(2,1,figsize=(10,6),sharex=True)
ax1.plot(pts['t'], pts['T'], 'b', linewidth=2)
ax1.set_ylabel('Temperature',fontsize=15)
ax2.plot(pts['t'], pts['A'], 'r', linewidth=2)
ax2.set_ylabel('Vegetation cover',fontsize=15)
ax2.set_xlabel('Time',fontsize=15)
#%%
# --------------------------------------------------------------------------- #
# SETUP CONTINUATION CLASS
# --------------------------------------------------------------------------- #

PC = ContClass(ode)
PCargs = args(name='EQ1', type='EP-C')
PCargs.freepars = ['gmma']
PCargs.StepSize = 1e-4
PCargs.MaxNumPoints = 130
PCargs.MaxStepSize = 1e-1
PCargs.MinStepSize = 1e-4
PCargs.LocBifPoints = 'all'
PCargs.verbosity = 2
PCargs.SaveEigen = True
PCargs.SaveJacobian = True
PC.newCurve(PCargs)

print('Computing curve...')
start = clock()
PC['EQ1'].backward()
print('done in %.3f seconds!' % (clock()-start))

PCargs = args(name='EQ2', type='EP-C')
PC.initpoint = 'EQ1:P1'
PCargs.freepars = ['gmma']
PCargs.StepSize = 1e-4
PCargs.MaxNumPoints = 300
PCargs.MaxStepSize = 1e-1
PCargs.MinStepSize = 1e-4
PCargs.LocBifPoints = 'all'
PCargs.verbosity = 2
PCargs.SaveEigen = True
PCargs.SaveJacobian = True
PC.newCurve(PCargs)

print('Computing curve...')
start = clock()
PC['EQ2'].forward()
print('done in %.3f seconds!' % (clock()-start))

#%%
# --------------------------------------------------------------------------- #
# PLOT BIFURCATION DIAGRAM
# --------------------------------------------------------------------------- #

PC['EQ1'].cleanLabels()
PC['EQ2'].cleanLabels()
a=plt.figure()
PC.display(('gmma','A'),stability=True,figure=a,linewidth=2)
plt.ylabel(r'Vegetation Cover $A$',fontsize=15)
plt.xlabel(r'Vegetation death rate $\gamma$',fontsize=15)
plt.title('Rombouts Vegetation model')
plt.show()
a=plt.figure()
PC.display(('gmma','T'),stability=True,figure=a,linewidth=2)
plt.ylabel(r'Temperature $T$',fontsize=15)
plt.xlabel(r'Vegetation death rate $\gamma$',fontsize=15)
plt.title('Rombouts Vegetation model')
plt.show()

File: Rombouts/Rombouts.f90
!----------------------------------------------------------------------
!----------------------------------------------------------------------
!   Rombouts Vegetation model
!----------------------------------------------------------------------
!----------------------------------------------------------------------

      SUBROUTINE FUNC(NDIM,U,ICP,PAR,IJAC,F,DFDU,DFDP)
!     ---------- ----


      IMPLICIT NONE
      INTEGER NDIM, IJAC, ICP(*)
      DOUBLE PRECISION U(NDIM), PAR(*), F(NDIM), DFDU(*), DFDP(*)
      DOUBLE PRECISION U1,U2, beta,alpha,R0,alpha0

	U1 = U(1)
	U2 = U(2)

	alpha0 	= MAX(PAR(6)+MIN(0.,(PAR(7)-PAR(6))*(U1-PAR(8))/(PAR(9)-PAR(8))),PAR(7))
	alpha 	= (1-PAR(3))*alpha0+PAR(3)*(PAR(4)*U2+PAR(5)*(1-U2))
	beta	= MAX(0.,1-PAR(13)*(U1-PAR(12))**2)
	R0	= PAR(10) + PAR(11)*(U1 - PAR(12))

	F(1) = ((1 - alpha)*PAR(2) - R0)/PAR(1)
	F(2) = beta*U2*(1 - U2) - PAR(14)*U2

      END SUBROUTINE FUNC
!----------------------------------------------------------------------
!----------------------------------------------------------------------

      SUBROUTINE STPNT(NDIM,U,PAR,T)

      IMPLICIT NONE
      INTEGER NDIM
      DOUBLE PRECISION U(NDIM), PAR(*), T

! Initialize the equation parameters
       PAR(1)=500.0	!Ct
       PAR(2)=342.5	!Q0
       PAR(3)=0.3	!p
       PAR(4)=0.1	!av
       PAR(5)=0.4	!ag
       PAR(6)=0.85	!alphamax
       PAR(7)=0.25	!alphamin
       PAR(8)=263.0	!Tal
       PAR(9)=300.0	!Tau
       PAR(10)=200.0	!B0
       PAR(11)=2.5	!B1
       PAR(12)=283.0	!Topt
       PAR(13)=0.004	!k
       PAR(14)=3.4608467752E-01!0.41777072213	!gmma

! Initialize the solution
       U(1)= 2.9432679889E+02!290.761399859
       U(2)= 2.8908303020E-01!0.449608397214

      END SUBROUTINE STPNT
!----------------------------------------------------------------------
!----------------------------------------------------------------------
      SUBROUTINE BCND 
      END SUBROUTINE BCND

      SUBROUTINE ICND 
      END SUBROUTINE ICND

      SUBROUTINE FOPT 
      END SUBROUTINE FOPT

      SUBROUTINE PVLS
      END SUBROUTINE PVLS
!----------------------------------------------------------------------
!----------------------------------------------------------------------
